instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java

```diff
@@ -0,0 +1,221 @@
+  private Worker worker;
+
+  private JarAnalyzer() {}
+
+  /** Get the {@link JarAnalyzer} singleton. */
+  public static JarAnalyzer getInstance() {
+    return INSTANCE;
+  }
+
+  /**
+   * Configure the {@link JarAnalyzer}. If not called, {@link #handle(ProtectionDomain)} is a noop.
+   *
+   * @throws IllegalStateException if called multiple times
+   */
+  public synchronized void configure(int jarsPerSecond) {
+    if (this.jarsPerSecond != null) {
+      throw new IllegalStateException("JarAnalyzer has already been configured");
+    }
+    this.jarsPerSecond = jarsPerSecond;
+    this.handler = this::handleInternal;
+  }
+
+  /**
+   * Install {@link OpenTelemetry} and start processing jars if {@link #configure(int)} was called.
+   *
+   * @throws IllegalStateException if called multiple times
+   */
+  public synchronized void maybeInstall(OpenTelemetry unused) {
+    if (worker != null) {
+      throw new IllegalStateException("JarAnalyzer has already been installed");
+    }
+    if (this.jarsPerSecond == null) {
+      return;
+    }
+    // TODO(jack-berg): Use OpenTelemetry to obtain EventEmitter when event API is stable
+    EventEmitter eventEmitter =
+        GlobalEventEmitterProvider.get()
+            .eventEmitterBuilder(JmxRuntimeMetricsUtil.getInstrumentationName())
+            .setInstrumentationVersion(JmxRuntimeMetricsUtil.getInstrumentationVersion())
+            .setEventDomain("jvm")
+            .build();
+    this.worker = new Worker(eventEmitter, toProcess, jarsPerSecond);
+    Thread workerThread =
+        new DaemonThreadFactory(JarAnalyzer.class.getSimpleName() + "_WorkerThread")
+            .newThread(worker);
+    workerThread.start();
+  }
+
+  /**
+   * Identify the JAR associated with the {@code protectionDomain} and queue it to be processed if
+   * its the first time we've seen it.
+   *
+   * <p>NOTE: does nothing if {@link #configure(int)} has not been called.
+   */
+  void handle(ProtectionDomain protectionDomain) {
+    this.handler.accept(protectionDomain);
+  }
+
+  private void handleInternal(ProtectionDomain protectionDomain) {
+    if (protectionDomain == null) {
+      return;
+    }
+    CodeSource codeSource = protectionDomain.getCodeSource();
+    if (codeSource == null) {
+      return;
+    }
+    URL jarUrl = codeSource.getLocation();
+    if (jarUrl == null) {
+      return;
+    }
+    URI locationUri;
+    try {
+      locationUri = jarUrl.toURI();
+    } catch (URISyntaxException e) {
+      LOGGER.log(Level.WARNING, "Unable to get URI for jar URL: " + jarUrl, e);
+      return;
+    }
+
+    if (!seenUris.add(locationUri)) {
+      return;
+    }
+    if ("jrt".equals(jarUrl.getProtocol())) {
+      LOGGER.log(Level.FINEST, "Skipping processing jar for java runtime module: " + jarUrl);
+      return;
+    }
+    if (!jarUrl.getFile().endsWith(JarAnalyzerUtil.JAR_EXTENSION)) {
+      LOGGER.log(Level.INFO, "Skipping processing jar with unrecognized code location: " + jarUrl);
```

Comment:
No probably not. My thought was we use logging to find all the shortcomings / edge cases with the current process to find / extract metadata. It will be hard to anticipate all the scenarios we'll come across in the wild so just trying to emphasize safety and good logging so we can improve the logic when we encounter issues.

Reply:
There are no more instances of this message being logged in the smoke tests. What log level do you think is appropriate such that we can discover edge cases with the logic without being noisy for users who don't care?

Reply:
I think info is reasonable for now. Feel free to re-open if you think otherwise.